payments-core/src/main/java/com/stripe/android/payments/core/authentication/DefaultPaymentAuthenticatorRegistry.kt

```diff
@@ -17,47 +18,63 @@ import javax.inject.Inject
 import kotlin.coroutines.CoroutineContext
 
 /**
- * Default registry to provide look ups for [IntentAuthenticator].
- * Should be only accessed through [DefaultIntentAuthenticatorRegistry.createInstance].
+ * Default registry to provide look ups for [PaymentAuthenticator].
+ * Should be only accessed through [DefaultPaymentAuthenticatorRegistry.createInstance].
  */
-internal class DefaultIntentAuthenticatorRegistry @Inject internal constructor(
+internal class DefaultPaymentAuthenticatorRegistry @Inject internal constructor(
     private val noOpIntentAuthenticator: NoOpIntentAuthenticator,
+    private val sourceAuthenticator: SourceAuthenticator,
     @IntentAuthenticatorMap
-    private val intentAuthenticatorMap:
-        Map<Class<out StripeIntent.NextActionData>, @JvmSuppressWildcards IntentAuthenticator>
-) : IntentAuthenticatorRegistry {
+    private val paymentAuthenticatorMap:
+        Map<Class<out StripeIntent.NextActionData>,
+            @JvmSuppressWildcards PaymentAuthenticator<StripeIntent>>
+) : PaymentAuthenticatorRegistry {
 
-    override fun getAuthenticator(stripeIntent: StripeIntent): IntentAuthenticator {
-        if (!stripeIntent.requiresAction()) {
-            return noOpIntentAuthenticator
-        }
-
-        return stripeIntent.nextActionData?.let {
-            intentAuthenticatorMap
-                .getOrElse(it::class.java) { noOpIntentAuthenticator }
-        } ?: run {
-            noOpIntentAuthenticator
+    @Suppress("UNCHECKED_CAST")
+    override fun <Authenticatable> getAuthenticator(
+        authenticatable: Authenticatable
+    ): PaymentAuthenticator<Authenticatable> {
+        return when (authenticatable) {
+            is StripeIntent -> {
+                if (!authenticatable.requiresAction()) {
+                    return noOpIntentAuthenticator as PaymentAuthenticator<Authenticatable>
+                }
+                return (
+                    authenticatable.nextActionData?.let {
+                        paymentAuthenticatorMap
+                            .getOrElse(it::class.java) { noOpIntentAuthenticator }
+                    } ?: run {
```

Comment:
Is this `run` needed?

Reply:
Yes, it's handles the case when `authenticatable.nextActionData` is null, without `run` it's will be checking the nullability of the entire block
```
authenticatable.nextActionData?.let {
    paymentAuthenticatorMap
        .getOrElse(it::class.java) { noOpIntentAuthenticator }
}
```

and throws this exception:
```
java.lang.ClassCastException: 
class com.stripe.android.payments.core.authentication.DefaultPaymentAuthenticatorRegistry$getAuthenticator$2 
cannot be cast to class com.stripe.android.payments.core.authentication.PaymentAuthenticator 
```